includes/class-amp-theme-support.php

```diff
@@ -14,6 +14,83 @@ class AMP_Theme_Support {
 
 	const COMPONENT_SCRIPTS_PLACEHOLDER = '<!--AMP_COMPONENT_SCRIPTS_PLACEHOLDER-->';
 
+	/**
+	 * Template types.
+	 *
+	 * @var array
+	 */
+	protected static $template_types = array(
+		'paged', // Deprecated.
+		'index',
+		'404',
+		'archive',
+		'author',
+		'category',
+		'tag',
+		'taxonomy',
+		'date',
+		'home',
+		'front_page',
+		'page',
+		'search',
+		'single',
+		'embed',
+		'singular',
+		'attachment',
+	);
+
+	/**
+	 * Initialize.
+	 */
+	public static function init() {
+		require_once AMP__DIR__ . '/includes/amp-post-template-actions.php';
+		if ( amp_is_canonical() ) {
+			$is_amp_endpoint = ( false !== get_query_var( AMP_QUERY_VAR, false ) ); // Because is_amp_endpoint() now returns true if amp_is_canonical().
+
+			// Permanently redirect to canonical URL if the AMP URL was loaded, since canonical is now AMP.
+			if ( $is_amp_endpoint ) {
+				wp_safe_redirect( self::get_current_canonical_url(), 301 );
+				exit;
+			}
+		} else {
+			self::register_paired_hooks();
+		}
+		self::register_hooks();
+	}
+
+	/**
+	 * Determines whether paired mode is available.
+	 *
+	 * When 'amp' theme support has not been added or canonical mode is enabled, then this returns false.
+	 * Returns true when there is a template_dir defined in theme support, and if a defined available_callback
+	 * returns true.
+	 *
+	 * @return bool Whether available.
+	 */
+	public static function is_paired_available() {
+		$support = get_theme_support( 'amp' );
+		if ( empty( $support ) || amp_is_canonical() ) {
+			return false;
+		}
+
+		$args = array_shift( $support );
+
+		if ( isset( $args['available_callback'] ) && is_callable( $args['available_callback'] ) ) {
+			return $args['available_callback']();
+		}
+		return true;
+	}
+
+	/**
+	 * Register hooks for paired mode.
+	 */
+	public static function register_paired_hooks() {
+		foreach ( self::$template_types as $template_type ) {
+			add_filter( "{$template_type}_template_hierarchy", array( __CLASS__, 'filter_paired_template_hierarchy' ) );
```

Comment:
It would be preferred if there was a `template_hierarchy` filter in core so that the list of types wouldn't have to be hard-coded. But alas, there is no such filter: https://github.com/WordPress/wordpress-develop/blob/fb5b6170564b403944a8fbf8a4e6eb877b26a092/src/wp-includes/template.php#L30-L44